Compiler selection changes

antico5 opened the pull request:
This PR introduces a new flag for solidity build profiles, `preferWasm`. When set to true, the wasm compiler will be used over the native one. All profiles have this set to `false` as default, except `production` which have this set to `true` by default.

Additionally, support for linux-aarch64 native compiler has been added. Some changes to the downloaded have been introduced since the source for the binaries doesn't behave like the official mirror for solc binaries. For instance, `keccak` checksum has been replaced with `sha256` to support both sources.

Since this is the only special case added, no design changes  have been made to the downloaded, although if we had more special cases in the future, we may need to think about using a strategy pattern or something similar.

Closes #5993

fvictorio commented on .changeset/nine-lies-press.md:
```diff
@@ -0,0 +1,7 @@
+---
+"@nomicfoundation/hardhat-verify": patch
```
Only the tests were changed in `hardhat-verify`, it shouldn't be included in the changeset.

fvictorio commented on v-next/hardhat/src/internal/builtin-plugins/solidity/build-system/build-system.ts:
```diff
@@ -506,6 +496,23 @@ export class SolidityBuildSystemImplementation implements SolidityBuildSystem {
   }
 
+  #getBuildProfile(_buildProfileName?: string) {
+    const buildProfileName = _buildProfileName ?? DEFAULT_BUILD_PROFILE;
```
Personal preference, I would replace these two lines with:

```typescript
  #getBuildProfile(buildProfileName: string = DEFAULT_BUILD_PROFILE) {
```

fvictorio commented on v-next/hardhat/src/internal/builtin-plugins/solidity/build-system/compiler/index.ts:
```diff
@@ -77,17 +77,22 @@ export async function downloadConfiguredCompilers(
-  );
+export async function getCompiler(
+  version: string,
+  preferWasm: boolean,
```
Can we make this an options object? Otherwise the calls to it look like this:

```ts
getCompiler(version, false)
```

instead of the much clearer:

```ts
getCompiler(version, { preferWasm: false })
```

fvictorio commented on v-next/hardhat/src/internal/builtin-plugins/solidity/build-system/compiler/index.ts:
```diff
@@ -77,17 +77,22 @@ export async function downloadConfiguredCompilers(
+  version: string,
+  preferWasm: boolean,
+): Promise<Compiler> {
+  if (preferWasm === false) {
```
`preferWasm` is not nullable nor optional, so `!preferWasm` should be fine here.

fvictorio commented on v-next/hardhat/src/internal/builtin-plugins/solidity/config.ts:
```diff
@@ -300,12 +309,17 @@ function resolveSolidityConfig(
     if (!(profile in profiles)) {
-      profiles[profile] = profiles.default;
+      profiles[profile] = {
+        ...profiles.default,
```
Unrelated to this particular PR, but we should probably do a deep copy here. Otherwise this happens:

<img width="620" height="180" alt="Image" src="https://github.com/user-attachments/assets/0e511be0-2da3-43e7-900d-f54fec26f9b2" />

It seems pedantic, but we've been bitten by this before.

But let's not do it in this PR. We should have a separate issue to make sure that both the user and the resolved config are made up of independent objects.

alcuadrado commented on v-next/hardhat/src/internal/builtin-plugins/solidity/build-system/build-system.ts:
```diff
@@ -343,7 +333,9 @@ export class SolidityBuildSystemImplementation implements SolidityBuildSystem {
       if (solcLongVersion === undefined) {
-        const compiler = await getCompiler(solcConfig.version);
+        const compiler = await getCompiler(solcConfig.version, {
+          preferWasm: false,
```
`{ preferWasm: buildProfile.preferWasm }`

This fixes it

alcuadrado replied:
It also was messing up with the cache system, as switching between wasm and this other build didn't generate different build ids.

antico5 replied:
Huge find! thanks for spotting it

alcuadrado reviewed (changes requested):
I left a few comments about two bugs I found when testing this on Linux aarch64.

I think we should include a `wasm: boolean` to `CompileCacheEntry`. This way we ensure that production builds don't get messed with the rest.

antico5 commented on v-next/hardhat/src/internal/builtin-plugins/solidity/build-system/compiler/downloader.ts:
```diff
@@ -354,12 +389,12 @@ export class CompilerDownloaderImplementation implements CompilerDownloader {
+    const compilerSha = bytesToHexString(await sha256(compiler));
 
-    if (expectedKeccak256 !== compilerKeccak256) {
+    if (expectedSha !== compilerSha) {
```
good catch

antico5 commented:
> I left a few comments about two bugs I found when testing this on Linux aarch64.
> 
> I think we should include a `wasm: boolean` to `CompileCacheEntry`. This way we ensure that production builds don't get messed with the rest.

Great catches. I will add wasm flag to compile cache entry. I didn't expect them to have the same long version as native compilers, but just checked and that is the case.

antico5 commented:
@alcuadrado Addressed the feedback, although I found that wasm is not working on windows (unrelated to these changes), but I'll fix it on this same PR
